qiskit_nature/second_q/operators/polynomial_tensor.py

```diff
@@ -0,0 +1,126 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Polynomial Tensor class"""
+
+from __future__ import annotations
+from typing import Dict, Mapping
+from numbers import Number
+import numpy as np
+from qiskit.quantum_info.operators.mixins import (
+    LinearMixin,
+    AdjointMixin,
+    TolerancesMixin,
+)
+
+
+class PolynomialTensor(LinearMixin, AdjointMixin, TolerancesMixin):
+    """Polynomial Tensor class"""
+
+    def __init__(self, data: Mapping[str, np.ndarray | Number]):
+        copy_dict: Dict[str, np.ndarray] = {}
+
+        shapes = set()
+        for key, value in data.items():
+            if isinstance(value, Number):
+                value = np.asarray(value)
+
+            if len(value.shape) != len(key):
+                raise ValueError(
+                    f"Data key {key} of length {len(key)} does not match "
+                    f"data value matrix of dimensions {value.shape}"
+                )
+
+            dims = set(value.shape)
+
+            if len(dims) > 1:
+                raise ValueError(
+                    f"For key {key}: dimensions of value matrix are not identical {value.shape}"
+                )
+
+            shapes.update(dims)
+            copy_dict[key] = value
+
+        if len(shapes) != 1:
+            raise ValueError("Dimensions of value matrices in data dictionary are not identical.")
+
+        self._data = copy_dict
+
+    def mul(self, other: complex):
+        """Scalar multiplication of PolynomialTensor with complex"""
+
+        if not isinstance(other, Number):
+            raise TypeError(f"other {other} must be a number")
+
+        prod_dict: Dict[str, np.ndarray] = {}
+        for key, matrix in self._data.items():
+            prod_dict[key] = np.multiply(matrix, other)
+        return PolynomialTensor(prod_dict)
+
+    def _multiply(self, other):
+        return self.mul(other)
+
+    def add(self, other: PolynomialTensor):
+        """Addition of PolynomialTensors"""
+
+        if not isinstance(other, PolynomialTensor):
+            raise TypeError("Incorrect argument type: other should be PolynomialTensor")
+
+        sum_dict: Dict[str, np.ndarray] = self._data.copy()
+        for other_key, other_value in other._data.items():
+            if other_key in sum_dict.keys():
+                if other_value.shape == np.shape(sum_dict[other_key]):
+                    sum_dict[other_key] = np.add(other_value, sum_dict[other_key])
+                else:
+                    raise ValueError(
+                        f"For key {other_key}: "
+                        f"corresponding data value of shape {np.shape(sum_dict[other_key])} "
+                        f"does not match other value matrix of shape {other_value.shape}"
+                    )
+            else:
+                sum_dict[other_key] = other_value
```

Comment:
I think we might be able to implement this nicer if we:
- have a `register_length` attribute which we can check once. This will then tell us that all shapes are correct
- then we can use dictionary comprehensions based on the assumptions that all shapes match

Again, you can see how we did this in the `SparseLabelOp` PR :+1: